### PR TITLE
MCCFR Enhancements and Bug Squashing

### PR DESCRIPTION
**Pull request by big-c-note**
Opening this branch to show progress on a few MCCFR enhancements and bug squashes. I'll comment when I'm in a stopping place for merging.

### CODE REVIEW

**Comment by big-c-note** [comment]
I wrapped up some logging changes into this PR. So far, in addition:

- An example of a proof of concept for saving memory by not holding on to sigma, as it can be calculated directly via regret. 
- An algorithm for creating an average offline strategy based off of regret dumps. We only need to dump regret, assuming that the training doesn't fail, the strategy is accurate as is. I guess it can't hurt to dump it just in case, but it does not need to be averaged.

**Comment by big-c-note** [comment]
Adding the changes to the main offline strategy algorithm. Not tested yet.

**Comment by big-c-note** [comment]
Using ```cmd + f DEBUG:root:Got EV``` between the two text files here you will see regret has not changed. The iterations however seem to be a little slower. Perhaps deleting the dictionary is not the best way. I will try just rewriting a variable to test the speed up.

**Comment by big-c-note** [comment]
Meant to add this file at commit ```a0c15d8```

**Comment by big-c-note** [comment]
Much faster implementation of the deleting sigma concept. The dictionary deletion was averaging ~45s/it on my machine, but the variable overwrite is averaging ~30s/it. Added example output to show regret is the same. 

I just need to confirm that a variable doesn't get altered once the algo returns to that level of recursion. For example, when updating regret for the traversing player, sigma would be updated by the opponents at various levels of recursion.

**Comment by big-c-note** [comment]
Double check me on the game logic for the test. I believe this is highlighting the bug I was showing you @fedden

**Comment by big-c-note** [comment]
Of course this PR is going to get an X because I wrote a test to show a bug.

**Comment by big-c-note** [comment]
Using cmd + f DEBUG:root:Got EV between the this and the original example files here you will see regret has not changed.

**Comment by big-c-note** [comment]
This is good to go

File: research/blueprint_algo/blueprint_short_deck_poker.py [hunk 1]
```diff
@@ -153,12 +128,14 @@ def calculate_strategy(
...
+    sigma = collections.defaultdict(lambda: collections.defaultdict(lambda: 1 / 3))
     rsum = sum([max(x, 0) for x in regret[I].values()])
     for a in state.legal_actions:
         if rsum > 0:
-            sigma[t + 1][I][a] = max(regret[I][a], 0) / rsum
+            sigma[I][a] = max(regret[I][a], 0) / rsum
         else:
-            sigma[t + 1][I][a] = 1 / len(state.legal_actions)
+            sigma[I][a] = 1 / len(state.legal_actions)
+    return sigma
```
Comment (fedden) [inline research/blueprint_algo/blueprint_short_deck_poker.py:138]: nice

File: test/functional/test_short_deck.py [hunk 2]
```diff
@@ -170,3 +171,52 @@ def test_pre_flop_pot(n_players: int, small_blind: int, big_blind: int):
     assert (
         n_bet_chips == pot.total
     ), f"small and big blind have are not in pot! {n_bet_chips} == {pot.total}"
+
+
+def _play_game_helper(state, betting_round_dict, bad_seq):
```
Comment (fedden) [inline test/functional/test_short_deck.py:176]: as a habit if you can leave a 1 liner docstring here just to describe at a glance what this does that would be awesome

Comment (fedden) [reply test/functional/test_short_deck.py:176]: if you wanna document the parameters that's always a bonus!

File: test/functional/test_short_deck.py [hunk 3]
```diff
@@ -170,3 +171,52 @@ def test_pre_flop_pot(n_players: int, small_blind: int, big_blind: int):
...
+    p = 1 / len(state.legal_actions)
+    probabilities = np.full(len(state.legal_actions), p)
+    a = np.random.choice(state.legal_actions, p=probabilities)
+    betting_stage = state._betting_stage
+    if betting_stage not in ["show_down", "terminal"]:
```
Comment (fedden) [inline test/functional/test_short_deck.py:181]: In future (not a big deal here as just 2 els), prefer a set `{"show_down", "terminal"}` due to O(1) lookup time

File: test/functional/test_short_deck.py [hunk 4]
```diff
@@ -170,3 +171,52 @@ def test_pre_flop_pot(n_players: int, small_blind: int, big_blind: int):
...
+        load_pickle_files=False,
+        small_blind=small_blind,
+        big_blind=big_blind,
+    )
+
+    # example of a bad sequence in a two-handed game in one round
+    bad_seq = ["raise", "call", "call"]
+    for t in range(100):
+        betting_round_dict = {"pre_flop": [], "flop": [], "turn": [], "river": []}
+        _play_game_helper(state, betting_round_dict, bad_seq)
```
Comment (fedden) [inline test/functional/test_short_deck.py:222]: This looks great. Have you tried hard-coding the specific example that you came across to catch the problem?

Comment (fedden) [reply test/functional/test_short_deck.py:222]: Well in fairness actually this test is picking up the error so this is a much more general test: https://travis-ci.org/github/fedden/poker_ai/builds/679596087

File: test/functional/test_short_deck.py (hunk 2 above)
Comment (big-c-note) [reply test/functional/test_short_deck.py:176]: Sounds good

File: test/functional/test_short_deck.py (hunk 3 above)
Comment (big-c-note) [reply test/functional/test_short_deck.py:181]: Good call, thanks for that

File: test/functional/test_short_deck.py (hunk 4 above)
Comment (big-c-note) [reply test/functional/test_short_deck.py:222]: I didn't hard code the specific bug, but added a test to catch the more general problem. The idea of adding a test was new to me to be honest, so feel free to rewrite if needed or let me know what should be changed

Comment (fedden) [reply test/functional/test_short_deck.py:222]: No as I said this is much more general than the hardcoded test I proposed

Comment (fedden) [reply test/functional/test_short_deck.py:222]: I much prefer this!

Comment (big-c-note) [reply test/functional/test_short_deck.py:222]: Ok great

**Pull request merged**